buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java

```diff
@@ -0,0 +1,165 @@
+
+    private static final Pattern JDK_VERSION = Pattern.compile("(\\d+)(\\.\\d+\\.\\d+)?\\+(\\d+)");
+
+    private List<String> ALLOWED_PLATFORMS = Collections.unmodifiableList(Arrays.asList("linux", "windows", "darwin"));
+
+    @Override
+    public void apply(Project project) {
+        NamedDomainObjectContainer<Jdk> jdksContainer = project.container(Jdk.class, name ->
+            new Jdk(name, project)
+        );
+        project.getExtensions().add("jdks", jdksContainer);
+
+        project.afterEvaluate(p -> {
+            for (Jdk jdk : jdksContainer) {
+                jdk.finalizeValues();
+                String version = jdk.getVersion();
+                String platform = jdk.getPlatform();
+
+                // depend on the jdk directory "artifact" from the root project
+                DependencyHandler dependencies = project.getDependencies();
+                Map<String, Object> depConfig = new HashMap<>();
+                depConfig.put("path", ":"); // root project
+                depConfig.put("configuration", configName("extracted_jdk", version, platform));
+                dependencies.add(jdk.getConfiguration().getName(), dependencies.project(depConfig));
+
+                // ensure a root level jdk download task exists
+                setupRootJdkDownload(project.getRootProject(), platform, version);
+            }
+        });
+    }
+
+    private static void setupRootJdkDownload(Project rootProject, String platform, String version) {
+        String extractTaskName = "extract" + capitalize(platform) + "Jdk" + version;
+        // NOTE: this is *horrendous*, but seems to be the only way to check for the existence of a registered task
+        try {
+            rootProject.getTasks().named(extractTaskName);
+            // already setup this version
+            return;
+        } catch (UnknownTaskException e) {
+            // fall through: register the task
+        }
+
+        // decompose the bundled jdk version, broken into elements as: [feature, interim, update, build]
+        // Note the "patch" version is not yet handled here, as it has not yet been used by java.
+        Matcher jdkVersionMatcher = Jdk.VERSION_PATTERN.matcher(version);
+        if (jdkVersionMatcher.matches() == false) {
+            throw new IllegalArgumentException("Malformed jdk version [" + version + "]");
+        }
+        String jdkVersion = jdkVersionMatcher.group(1) + (jdkVersionMatcher.group(2) != null ? (jdkVersionMatcher.group(2)) : "");
+        String jdkMajor = jdkVersionMatcher.group(1);
+        String jdkBuild = jdkVersionMatcher.group(3);
+
+        // add fake ivy repo for jdk url
+        String repoName = "jdk_repo_" + version;
+        if (rootProject.getRepositories().findByName(repoName) == null) {
+            rootProject.getRepositories().ivy(ivyRepo -> {
+                ivyRepo.setName(repoName);
+                ivyRepo.setUrl("https://download.java.net");
```

Comment:
This would add a considerable load to our infrastructure. We may want to still consider it, but I think it should be separate from this PR, which is just to consolidate existing code, which already uses this url.

Reply:
I still have a bit of a concern with this approach of registering so many repositories. Is there anyway perhaps that we could construct the dependency notation so that we can avoid creating a repo per version? I'm not sure this is possible, or if what I see here even works. For example, the download link I see [here](https://jdk.java.net/12/) includes the artifact hash in the download URL.

Reply:
I don't think it is possible here. The reason we embed this information in the pattern is we only have so many elements we can substitute via the maven/ivy coordinates. 

The hash included in the download link is something that changed very recently. There are 2 versions registered here because versions before 12.0.1 still use the old pattern. I could definitely decide which one to register based on the version, but in this PR I'm only trying to move the logic we already have in the build for jdk downloading. We can certainly look at improving this either through reducing repos or moving to using the jvm catalog, but I think these should be discussed separately as followups to this PR.

Reply:
Since we're only registering all these repos to the root project it's probably ok, as it won't affect dependency solution of subprojects. Let's make sure we create an issue to follow up here. I assume if the recent version download links have changed this is going to be a problem for fetching newer JDK versions, no?

Reply:
The older links didn't change. Oracle only updated the url pattern for newer jdk releases. This is why we have both patterns here.

Reply:
I think I'm missing something. I only see a single artifact pattern definition. Is this going to work with newer JDKs that require the artifact hash to be part of the URL?

Reply:
We register 2 ivy repositories, one with the older pattern, and one with the hash. Are you sure you are looking at the latest commit?

Reply:
I definitely was not. It seems GitHub was filtering out the latest commits for me. Thanks GH!

Reply:
Given that we know the version when registering the repo can we be smart about it and avoid registering both for all versions?

Reply:
Or simpler yet, if the artifact version notation includes a hash, use the hash repo, otherwise use the legacy one.

Reply:
Sure, I will do that in a followup.